File: app/src/main/java/com/example/a8_puzzle/solver/PuzzleSolver.java
package com.example.a8_puzzle.solver;

import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.PriorityQueue;

public class PuzzleSolver {
    
    private static final String TAG = "PuzzleSolver.java";
    private List<Puzzle> steps;
    private Node solvedNode;
    private boolean isSolvable;
    
    // find a solution using the A* algorithm
    public PuzzleSolver(Puzzle initial) {
        Node initialNode = new Node(initial, 0, null);
        
        // Java's PriorityQueue is a minimum one by default (ordered in ascending order)
        PriorityQueue<Node> solver = new PriorityQueue<>();
        
        // insert initial nodes
        solver.add(initialNode);
        solve(solver);
    }
    
    private void solve(PriorityQueue<Node> solver) {
        
        while (true) {
            // delete min from priority queue
            Node delMin = solver.poll();
    
            // if deleted node is the solved puzzle, break
            if (delMin.puzzle.isSolved()) {
                solvedNode = delMin;
                isSolvable = true;
                return;
            }
            // insert all the neighbouring nodes to the deleted one
            for (Puzzle p : delMin.puzzle.neighbors()) {
                // do not insert the neighbour identical to deleted board's previous one
                if (delMin.prevNode == null || !p.equals(delMin.prevNode.puzzle)) {
                    solver.add(new Node(p, delMin.numOfMoves + 1, delMin));
                }
            }
        }
    }
    
    // min number of moves to solve initial board
    public int moves() {
        if (steps == null) {
            // if solution hasn't been called yet
            solution();
        }
        // then get size of steps
        if (isSolvable) {
            return steps.size() - 1;
            
        } else {
            return -1;
        }
    }
    
    // sequence of boards in a shortest solution
    public List<Puzzle> solution() {
        if (steps == null) {
            // if solution() not called yet
            steps = new ArrayList<>();
            if (isSolvable) {
                Node current = solvedNode;
                while (current != null) {
                    if (current.movement != null) {
                        current.puzzle.setMovement(current.movement);
                    } else {
                        // first puzzle (with no previous node, so no movement)
                        current.puzzle.setMovement("Starting puzzle: ");
                    }
                    steps.add(current.puzzle);
                    current = current.prevNode;
                }
                // steps were placed in reverse order (starting with the solved node)
                Collections.reverse(steps);
                return steps;
                
            } else {
                // if unsolvable return null
                return null;
            }
            
        } else {
            // else solution() has already been called; return the stored solution
            return steps;
        }
    }
    
    // create immutable object Node to represent each state of the puzzle tiles
    private static class Node implements Comparable<Node> {
        
        private final Puzzle puzzle;
        private final int numOfMoves;
        private final Node prevNode;
        private final int priority;
        private String movement;
    
    
        public Node(Puzzle puzzle, int numOfMoves, @Nullable Node prevNode) {
            this.puzzle = puzzle;
            this.numOfMoves = numOfMoves;
            this.prevNode = prevNode;
            // define the priority function as its manhattan distance plus the number of moves so far.
            priority = puzzle.manhattanDistance() + this.numOfMoves;
            if (prevNode != null) {
                movement = prevNode.puzzle.movement(this.puzzle);
            }
        }
        
        // implement Comparable methods for PriorityQueue
        @Override
        public int compareTo(Node that) {
            if (this.priority > that.priority) return +1;
            else if (this.priority < that.priority) return -1;
            else return Integer.compare(this.numOfMoves, that.numOfMoves);
        }
    }
}

File: app/src/main/java/com/example/cool8puzzle/solver/PuzzleSolver.kt
package com.example.cool8puzzle.solver

// find a solution using the A* algorithm

import com.example.cool8puzzle.entity.Puzzle
import java.util.PriorityQueue

class PuzzleSolver constructor(initial: Puzzle? = null) {

    private var steps: MutableList<Puzzle>? = null
    private var solvedNode: Node? = null
    private var isSolvable = false

    init {
        initial?.let { solve(it) }
    }

    fun solve(puzzle: Puzzle) {
        val initialNode = Node(puzzle, 0, null)

        // Java's PriorityQueue is a minimum one by default (ordered in ascending order)
        val solver = PriorityQueue<Node>()

        solver.add(initialNode)
        if (puzzle.isSolvable()) {
            solve(solver)
        }
    }

    private fun solve(solver: PriorityQueue<Node>) {
        while (true) {
            val delMin = solver.poll()

            if (delMin != null) {
                if (delMin.puzzle.isSolved()) {
                    solvedNode = delMin
                    isSolvable = true
                    return
                }

                for (p in delMin.puzzle.neighbors()) {
                    // do not insert the neighbour identical to deleted board's previous one
                    if (delMin.prevNode == null || p != delMin.prevNode.puzzle) {
                        solver.add(
                            Node(
                                p,
                                delMin.numOfMoves + 1,
                                delMin
                            )
                        )
                    }
                }
            }
        }
    }

    fun moves(): Int {
        if (steps == null) {
            solution()
        }

        return if (isSolvable) {
            steps!!.size - 1
        } else {
            -1
        }
    }

    fun solution(): MutableList<Puzzle>? {
        return if (steps == null) {
            steps = mutableListOf()
            if (isSolvable) {
                var current = solvedNode
                while (current != null) {
                    if (current.movement != null) {
                        // add the movement text
                        current.puzzle.movement = current.movement
                    } else {
                        // first puzzle (with no previous node, so no movement)
                        current.puzzle.movement = "Starting puzzle: "
                    }
                    steps!!.add(current.puzzle)
                    current = current.prevNode
                }
                steps!!.reverse() // steps were placed in reverse order
                steps
            } else {
                null
            }
        } else {
            // return the stored solution
            steps
        }
    }

    // create Node to represent each state of the puzzle tiles
    private class Node(
        val puzzle: Puzzle,
        val numOfMoves: Int,
        val prevNode: Node?
    ) : Comparable<Node> {

        // define the priority function as its manhattan distance plus the number of moves so far.
        private val priority: Int = puzzle.manhattanDistance() + numOfMoves

        var movement: String? = null

        override fun compareTo(other: Node): Int {
            return when {
                priority > other.priority -> 1
                priority < other.priority -> -1
                else -> numOfMoves.compareTo(other.numOfMoves)
            }
        }

        init {
            if (prevNode != null) {
                movement = prevNode.puzzle.movement(puzzle)
            }
        }
    }
}